src/main/scala/com/gu/invoicing/invoice/Impl.scala

```diff
@@ -0,0 +1,130 @@
+import java.time.LocalDate
+import com.gu.invoicing.invoice.Model._
+import scalaj.http.Http
+import scala.util.chaining._
+import com.gu.spy._
+
+/**
+ * Zuora API client and implementation details
+ */
+object Impl {
+  lazy val stage = getenv("Stage")
+
+  lazy val zuoraApiHost: String =
+    stage match { case "DEV" | "CODE" => "https://rest.apisandbox.zuora.com"; case "PROD" => "https://rest.zuora.com" }
+
+  lazy val config = read[Config](getenv("Config"))
+
+  lazy val accessToken: String = {
+    Http(s"$zuoraApiHost/oauth/token")
+      .postForm(Seq(
+        "client_id" -> config.clientId,
+        "client_secret" -> config.clientSecret,
+        "grant_type" -> "client_credentials"
+      ))
+      .asString
+      .body
+      .pipe(read[AccessToken](_))
+      .access_token
+  }
+
+  def getInvoices(account: String): List[Invoice] = {
+    Http(s"$zuoraApiHost/v1/transactions/invoices/accounts/$account")
+      .header("Authorization", s"Bearer ${accessToken}")
+      .asString
+      .body
+      .pipe(read[Invoices](_))
+      .invoices
+      .filter { _.amount >= 0.0 }
+  }
+
+  def getPayments(account: String): List[Payment] = {
+    Http(s"$zuoraApiHost/v1/transactions/payments/accounts/$account")
+      .header("Authorization", s"Bearer ${accessToken}")
+      .asString
+      .body
+      .pipe(read[Payments](_))
+      .payments
+  }
+
+  def getPaymentMethods(accountId: String): List[PaymentMethod] = {
+    Http(s"$zuoraApiHost/v1/action/query")
+      .header("Authorization", s"Bearer ${accessToken}")
+      .header("Content-Type", "application/json")
+      .postData(s"""{"queryString": "select BankTransferType, CreditCardExpirationMonth, CreditCardExpirationYear, BankTransferAccountNumberMask, LastFailedSaleTransactionDate, LastTransactionDateTime, LastTransactionStatus, Name, NumConsecutiveFailures, PaymentMethodStatus, Type, ID, MandateID, PaypalBAID, SecondTokenID, TokenID, AccountID, Active, Country, CreatedById, CreatedDate, CreditCardType, DeviceSessionId, IdentityNumber, MandateCreationDate, MandateReceived, MandateUpdateDate, MaxConsecutivePaymentFailures, PaymentRetryWindow, TotalNumberOfErrorPayments, TotalNumberOfProcessedPayments, UpdatedById, UpdatedDate, UseDefaultRetryRule, CreditCardMaskNumber from PaymentMethod where AccountId = '$accountId'"}""")
+      .method("POST")
+      .asString
+      .body
+      .pipe(read[PaymentMethods](_))
+      .records
+  }
+
+  def getPaymentMethod(
+    invoiceId: String,
+    invoices: List[Invoice],
+    payments: List[Payment],
+    paymentMethods: List[PaymentMethod]
+  ): PaymentMethod = {
+    val paymentMethodIdByInvoiceId: Map[String, String] =
+      payments
+        .flatMap { payment =>
+          payment
+            .paidInvoices
+            .tap(v => assert(v.length == 1, s"Payment should be associated with only one invoice: $payment"))
+            .headOption
+            .map { invoice => invoice.invoiceId -> payment.paymentMethodId }
```

Comment:
could be:
```suggestion
            .map { _.invoiceId -> payment.paymentMethodId }
```
but slight lack of clarity